File: lianel/dc/frontend/src/energy/Energy.js
import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import PageTemplate from '../PageTemplate';
import { energyApi } from './energyApi';
import { TimeSeriesChart, CountryComparisonChart, ProductDistributionChart, FlowDistributionChart } from './EnergyCharts';
import '../App.css';
import './Energy.css';

function Energy() {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');
  const [serviceInfo, setServiceInfo] = useState(null);
  const [energyData, setEnergyData] = useState(null);
  const [filters, setFilters] = useState({
    country_code: '',
    year: '',
    limit: 50,
    offset: 0
  });
  const [summary, setSummary] = useState(null);

  useEffect(() => {
    fetchData();
  }, [filters.offset]);

  const fetchData = async () => {
    try {
      setLoading(true);
      setError('');

      // Fetch service info
      const info = await energyApi.getServiceInfo();
      setServiceInfo(info);

      // Fetch energy data
      const params = {
        limit: parseInt(filters.limit) || 50,
        offset: filters.offset || 0
      };
      if (filters.country_code) params.country_code = filters.country_code;
      if (filters.year) params.year = parseInt(filters.year);

      const data = await energyApi.getEnergyAnnual(params);
      setEnergyData(data);

      // Fetch summary
      const summaryData = await energyApi.getEnergySummary({
        country_code: filters.country_code || undefined,
        year: filters.year ? parseInt(filters.year) : undefined,
        group_by: 'country'
      });
      setSummary(summaryData);
    } catch (err) {
      console.error('Error fetching energy data:', err);
      setError(err.message || 'Failed to load energy data');
    } finally {
      setLoading(false);
    }
  };

  const handleFilterChange = (field, value) => {
    setFilters(prev => ({ ...prev, [field]: value, offset: 0 }));
  };

  const handleApplyFilters = () => {
    fetchData();
  };

  const handleResetFilters = () => {
    setFilters({ country_code: '', year: '', limit: 50, offset: 0 });
    setTimeout(fetchData, 100);
  };

  const formatNumber = (num) => {
    if (num === null || num === undefined) return 'N/A';
    return new Intl.NumberFormat('en-US', {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2
    }).format(num);
  };

  // Calculate summary metrics
  const calculateMetrics = () => {
    if (!energyData || !energyData.data) return null;
    
    const data = energyData.data;
    const totalGWh = data.reduce((sum, record) => sum + (parseFloat(record.value_gwh) || 0), 0);
    const avgGWh = data.length > 0 ? totalGWh / data.length : 0;
    const countries = new Set(data.map(r => r.country_code)).size;
    const years = new Set(data.map(r => r.year)).size;
    
    return {
      totalGWh,
      avgGWh,
      countries,
      years,
      recordCount: data.length
    };
  };

  // Export to CSV
  const exportToCSV = () => {
    if (!energyData || !energyData.data || energyData.data.length === 0) {
      alert('No data to export');
      return;
    }

    const headers = ['Country Code', 'Country Name', 'Year', 'Product Code', 'Product Name', 'Flow Code', 'Flow Name', 'Value (GWh)', 'Unit', 'Source Table'];
    const rows = energyData.data.map(record => [
      record.country_code || '',
      record.country_name || '',
      record.year || '',
      record.product_code || '',
      record.product_name || '',
      record.flow_code || '',
      record.flow_name || '',
      record.value_gwh || '0',
      record.unit || '',
      record.source_table || ''
    ]);

    const csvContent = [
      headers.join(','),
      ...rows.map(row => row.map(cell => `"${String(cell).replace(/"/g, '""')}"`).join(','))
    ].join('\n');

    const blob = new Blob([csvContent], { type: 'text/csv;charset=utf-8;' });
    const link = document.createElement('a');
    const url = URL.createObjectURL(blob);
    link.setAttribute('href', url);
    link.setAttribute('download', `energy-data-${new Date().toISOString().split('T')[0]}.csv`);
    link.style.visibility = 'hidden';
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
  };

  const metrics = calculateMetrics();

  if (loading && !energyData) {
    return (
      <PageTemplate title="EU Energy Data">
        <div className="loading">Loading energy data...</div>
      </PageTemplate>
    );
  }

  return (
    <PageTemplate title="EU Energy Data">
      <div className="energy-container">
        {/* Summary Cards */}
        {serviceInfo && (
          <div className="summary-cards">
            <div className="summary-card">
              <div className="card-icon">📊</div>
              <div className="card-content">
                <div className="card-label">Total Records</div>
                <div className="card-value">{serviceInfo.database.total_records.toLocaleString()}</div>
              </div>
            </div>
            <div className="summary-card">
              <div className="card-icon">🌍</div>
              <div className="card-content">
                <div className="card-label">Countries</div>
                <div className="card-value">{serviceInfo.database.countries}</div>
              </div>
            </div>
            <div className="summary-card">
              <div className="card-icon">📅</div>
              <div className="card-content">
                <div className="card-label">Years</div>
                <div className="card-value">{serviceInfo.database.years}</div>
              </div>
            </div>
            {metrics && (
              <>
                <div className="summary-card">
                  <div className="card-icon">⚡</div>
                  <div className="card-content">
                    <div className="card-label">Total Energy (GWh)</div>
                    <div className="card-value">{formatNumber(metrics.totalGWh)}</div>
                  </div>
                </div>
                <div className="summary-card">
                  <div className="card-icon">📈</div>
                  <div className="card-content">
                    <div className="card-label">Avg per Record</div>
                    <div className="card-value">{formatNumber(metrics.avgGWh)}</div>
                  </div>
                </div>
              </>
            )}
          </div>
        )}

            {error && <div className="error-message">{error}</div>}

      {/* Filters */}
      <div className="filters-panel">
        <h3>Filters</h3>
        <div className="filters-grid">
          <div className="filter-group">
            <label>Country Code (e.g., SE, DE, FR):</label>
            <input
              type="text"
              value={filters.country_code}
              onChange={(e) => handleFilterChange('country_code', e.target.value.toUpperCase())}
              placeholder="SE"
              maxLength="2"
            />
          </div>
          <div className="filter-group">
            <label>Year:</label>
            <input
              type="number"
              value={filters.year}
              onChange={(e) => handleFilterChange('year', e.target.value)}
              placeholder="2023"
              min="2015"
              max="2024"
            />
          </div>
          <div className="filter-group">
            <label>Records per page:</label>
            <select
              value={filters.limit}
              onChange={(e) => handleFilterChange('limit', e.target.value)}
            >
              <option value="25">25</option>
              <option value="50">50</option>
              <option value="100">100</option>
              <option value="200">200</option>
            </select>
          </div>
        </div>
        <div className="filter-actions">
          <button onClick={handleApplyFilters} className="btn-primary">Apply Filters</button>
          <button onClick={handleResetFilters} className="btn-secondary">Reset</button>
        </div>
      </div>

      {/* Charts Section */}
      {energyData && energyData.data && energyData.data.length > 0 && (
        <div className="charts-section">
          <h2>Data Visualization</h2>
          <div className="charts-grid">
            <TimeSeriesChart data={energyData} countryCode={filters.country_code} />
            {summary && summary.summary && summary.summary.length > 0 && (
              <CountryComparisonChart summary={summary} />
            )}
            <ProductDistributionChart data={energyData} />
            <FlowDistributionChart data={energyData} />
          </div>
        </div>
      )}

      {/* Summary */}
      {summary && summary.summary && summary.summary.length > 0 && (
        <div className="summary-panel">
          <h3>Summary by Country</h3>
          <div className="summary-grid">
            {summary.summary.slice(0, 10).map((item, idx) => (
              <div key={idx} className="summary-item">
                <div className="summary-label">{item.group}</div>
                <div className="summary-value">{formatNumber(item.total_gwh)} GWh</div>
                <div className="summary-count">{item.record_count} records</div>
              </div>
            ))}
          </div>
        </div>
      )}

      {/* Data Table */}
      {energyData && (
        <div className="data-panel">
          <div className="data-panel-header">
            <h3>
              Energy Records
              {energyData.total !== undefined && (
                <span className="total-count"> ({energyData.total.toLocaleString()} total)</span>
              )}
            </h3>
            <button onClick={exportToCSV} className="btn-export">
              📥 Export to CSV
            </button>
          </div>
          <div className="table-container">
            <table className="energy-table">
              <thead>
                <tr>
                  <th>Country</th>
                  <th>Year</th>
                  <th>Product</th>
                  <th>Flow</th>
                  <th>Value (GWh)</th>
                  <th>Source</th>
                </tr>
              </thead>
              <tbody>
                {energyData.data && energyData.data.length > 0 ? (
                  energyData.data.map((record, idx) => (
                    <tr key={idx}>
                      <td>
                        <strong>{record.country_code}</strong>
                        {record.country_name && <div className="country-name">{record.country_name}</div>}
                      </td>
                      <td>{record.year}</td>
                      <td>
                        {record.product_code || 'N/A'}
                        {record.product_name && <div className="detail-text">{record.product_name}</div>}
                      </td>
                      <td>
                        {record.flow_code || 'N/A'}
                        {record.flow_name && <div className="detail-text">{record.flow_name}</div>}
                      </td>
                      <td className="value-cell">{formatNumber(record.value_gwh)}</td>
                      <td>{record.source_table}</td>
                    </tr>
                  ))
                ) : (
                  <tr>
                    <td colSpan="6" className="no-data">No data found</td>
                  </tr>
                )}
              </tbody>
            </table>
          </div>

          {/* Pagination */}
          {energyData.total > filters.limit && (
            <div className="pagination">
              <button
                onClick={() => setFilters(prev => ({ ...prev, offset: Math.max(0, prev.offset - prev.limit) }))}
                disabled={filters.offset === 0}
                className="btn-secondary"
              >
                Previous
              </button>
              <span className="page-info">
                Showing {filters.offset + 1} - {Math.min(filters.offset + filters.limit, energyData.total)} of {energyData.total}
              </span>
              <button
                onClick={() => setFilters(prev => ({ ...prev, offset: prev.offset + prev.limit }))}
                disabled={filters.offset + filters.limit >= energyData.total}
                className="btn-secondary"
              >
                Next
              </button>
            </div>
          )}
        </div>
      )}
      </div>
    </PageTemplate>
  );
}

export default Energy;

File: lianel/dc/energy-service/src/handlers/energy.rs
use axum::{
    extract::{Query, State, Path},
    response::Json,
    http::StatusCode,
};
use sqlx::PgPool;
use crate::models::*;
use crate::db::queries;

#[utoipa::path(
    get,
    path = "/api/energy/annual",
    tag = "energy",
    params(
        ("country_code" = Option<String>, Query, description = "Country code filter"),
        ("year" = Option<i32>, Query, description = "Year filter"),
        ("product_code" = Option<String>, Query, description = "Product code filter"),
        ("flow_code" = Option<String>, Query, description = "Flow code filter"),
        ("source_table" = Option<String>, Query, description = "Source table filter"),
        ("limit" = Option<u32>, Query, description = "Limit (default: 1000)"),
        ("offset" = Option<u32>, Query, description = "Offset (default: 0)"),
    ),
    responses(
        (status = 200, description = "Energy data retrieved successfully", body = EnergyResponse),
        (status = 500, description = "Internal server error")
    )
)]
pub async fn get_energy_annual(
    State(pool): State<PgPool>,
    Query(params): Query<EnergyQueryParams>,
) -> Result<Json<EnergyResponse>, (StatusCode, Json<serde_json::Value>)> {
    let limit = params.limit.unwrap_or(1000).min(10000); // Max 10k records
    let offset = params.offset.unwrap_or(0);

    match queries::get_energy_records(
        &pool,
        params.country_code.as_deref(),
        params.year,
        params.product_code.as_deref(),
        params.flow_code.as_deref(),
        params.source_table.as_deref(),
        limit,
        offset,
    ).await {
        Ok((records, total)) => Ok(Json(EnergyResponse {
            data: records,
            total,
            limit,
            offset,
        })),
        Err(e) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({
                "error": "Failed to query energy data",
                "details": e.to_string()
            })),
        )),
    }
}

#[utoipa::path(
    get,
    path = "/api/energy/annual/by-country/{country_code}",
    tag = "energy",
    params(
        ("country_code" = String, Path, description = "Country code"),
    ),
    responses(
        (status = 200, description = "Energy data retrieved successfully", body = EnergyResponse),
    )
)]
pub async fn get_energy_by_country(
    State(pool): State<PgPool>,
    Path(country_code): Path<String>,
    Query(params): Query<EnergyQueryParams>,
) -> Result<Json<EnergyResponse>, (StatusCode, Json<serde_json::Value>)> {
    let limit = params.limit.unwrap_or(1000).min(10000);
    let offset = params.offset.unwrap_or(0);

    match queries::get_energy_records(
        &pool,
        Some(&country_code),
        params.year,
        params.product_code.as_deref(),
        params.flow_code.as_deref(),
        params.source_table.as_deref(),
        limit,
        offset,
    ).await {
        Ok((records, total)) => Ok(Json(EnergyResponse {
            data: records,
            total,
            limit,
            offset,
        })),
        Err(e) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({
                "error": "Failed to query energy data",
                "details": e.to_string()
            })),
        )),
    }
}

#[utoipa::path(
    get,
    path = "/api/energy/annual/by-year/{year}",
    tag = "energy",
    params(
        ("year" = i32, Path, description = "Year"),
    ),
    responses(
        (status = 200, description = "Energy data retrieved successfully", body = EnergyResponse),
    )
)]
pub async fn get_energy_by_year(
    State(pool): State<PgPool>,
    Path(year): Path<i32>,
    Query(params): Query<EnergyQueryParams>,
) -> Result<Json<EnergyResponse>, (StatusCode, Json<serde_json::Value>)> {
    let limit = params.limit.unwrap_or(1000).min(10000);
    let offset = params.offset.unwrap_or(0);

    match queries::get_energy_records(
        &pool,
        params.country_code.as_deref(),
        Some(year),
        params.product_code.as_deref(),
        params.flow_code.as_deref(),
        params.source_table.as_deref(),
        limit,
        offset,
    ).await {
        Ok((records, total)) => Ok(Json(EnergyResponse {
            data: records,
            total,
            limit,
            offset,
        })),
        Err(e) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({
                "error": "Failed to query energy data",
                "details": e.to_string()
            })),
        )),
    }
}

#[utoipa::path(
    get,
    path = "/api/energy/annual/summary",
    tag = "energy",
    params(
        ("country_code" = Option<String>, Query, description = "Country code filter"),
        ("year" = Option<i32>, Query, description = "Year filter"),
        ("group_by" = Option<String>, Query, description = "Group by: country, year, product, flow (default: country)"),
    ),
    responses(
        (status = 200, description = "Summary retrieved successfully", body = EnergySummaryResponse),
    )
)]
pub async fn get_energy_summary(
    State(pool): State<PgPool>,
    Query(params): Query<serde_json::Value>,
) -> Result<Json<EnergySummaryResponse>, (StatusCode, Json<serde_json::Value>)> {
    let country_code = params.get("country_code")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());
    let year = params.get("year")
        .and_then(|v| v.as_i64())
        .map(|y| y as i32);
    let group_by = params.get("group_by")
        .and_then(|v| v.as_str())
        .unwrap_or("country");

    match queries::get_energy_summary(
        &pool,
        country_code.as_deref(),
        year,
        group_by,
    ).await {
        Ok(results) => {
            let summary: Vec<EnergySummary> = results
                .into_iter()
                .map(|(group, total_gwh, record_count)| EnergySummary {
                    group,
                    total_gwh,
                    record_count,
                })
                .collect();

            let total_records: i64 = summary.iter().map(|s| s.record_count).sum();

            Ok(Json(EnergySummaryResponse {
                summary,
                total_records,
            }))
        },
        Err(e) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(serde_json::json!({
                "error": "Failed to query summary",
                "details": e.to_string()
            })),
        )),
    }
}

